### ISSUE TITLE
Automatic Brace Insertion (ABI) is undesirable

### ISSUE THREAD

**Issue by robpalme** [microsoft/TypeScript#34876]
**TypeScript Version:**  3.7-beta

**Search Terms:** automatic brace insertion ABI syntax error

**Code:**

```ts
function a() {
    "Automatic Brace Insertion! (ABI)";
```

**Expected behavior:**

The input is syntactically invalid, so emit should fail regardless of the state of `noEmitOnError`.  Other forms of un-parsable input already do this, e.g. [`function a(`](https://www.typescriptlang.org/play/?alwaysStrict=false&ssl=3&ssc=2&pln=1&pc=1#code/GYVwdgxgLglg9mABAQwBSIN4ChG8QIgFUwoBTAJwFsYxkyATRAB2XNJIAtSBnHxCOOTbQANgE9EbUL26IocRKWpR+ceqQB0+ANxYAvkA)

**Actual behavior:**

A closing brace is inserted in the emitted JS code.

```js
function a() {
    "Automatic Brace Insertion! (ABI)";
}
```

This means that users who are not using `noEmitOnError` are silently having invalid code transformed into valid code.  I can confirm this actually happened to someone shipping something to production.

**Playground Link:** [ABI Playground Link](https://www.typescriptlang.org/play/?alwaysStrict=false&ssl=1&ssc=1&pln=2&pc=40#code/GYVwdgxgLglg9mABAQwBQEpEG8BQj+IBEAgiFHALbKwSIBCATshAKaICSYAziw7AgEJEqYnXbpCAbhw4gA)

**Related Issues:** https://github.com/microsoft/TypeScript/issues/34870 , https://github.com/microsoft/TypeScript/issues/34871 , [Lazy Twitter report](https://twitter.com/robpalmer2/status/1190383568612184064)

I would suggest this bug deserves to be fixed based on [Design Goal 7](https://github.com/microsoft/TypeScript/wiki/TypeScript-Design-Goals#goals) (_Preserve runtime behavior of all JavaScript code_) and [Design Non-Goal 7](https://github.com/microsoft/TypeScript/wiki/TypeScript-Design-Goals#non-goals) (_Introduce behaviour that is likely to surprise users_).

**Comment by fatcerberus** [comment]
IIRC, this is at least partially a design limitation insofar as TS doesn’t internally make enough distinction between syntax errors and semantic errors to selectively prevent emit only in the former case.  Braces are implicit in an AST structure, which means after parsing the code in the OP there’s enough information in the AST to emit — and so it does.

See comment here for more background on the above design limitation: https://github.com/microsoft/TypeScript/issues/28868#issuecomment-446792538

**Comment by fatcerberus** [comment]
> **Tagline:** *If the braces don't fit, you must not emit.*

😆

**Comment by RyanCavanaugh** [comment]
I really align with the spirit of this. Syntactic errors used to be uniformly produced during parsing, and semantic errors during checking, but over time we allowed certain ungrammatical things to be parsed and then have errors issued later during checking - doing so is sometimes very favorable in terms of the completion experience in the editor.

The work here is technically straightfoward (tag every error diagnostic as emit-blocking or not) but it's just a question of whether it's worth it to maintain and store that data. We also had some disagreements, e.g.
```ts
interface X extends { }
```
is certainly a syntactic error, but arguably shouldn't block emit? Or maybe it's syntactically valid but has a semantic error because the `extends` clause has the illegal length 0? The distinction is not always 100.0% apparent. Or maybe syntax errors in type positions shouldn't matter at all since they're getting erased anyway?

**Comment by robpalme** [comment]
Yes, I can appreciate the pay off for this work might seem low for the amount of effort/maintenance.  I won't argue this is high priority, given the workaround is to use `"noEmitOnError"`.

Maybe we could lower the effort by scoping this issue to a basic set of obvious errors, e.g. unbalanced braces/brackets.
